app/scripts/modules/core/pipeline/config/graph/pipeline.graph.component.ts

```diff
@@ -0,0 +1,377 @@
+import { debounce, filter, find, flatten, forOwn, groupBy, max, maxBy, sortBy, sum, sumBy, throttle, uniq } from 'lodash';
+import { module } from 'angular';
+
+import { IExecution, IPipeline } from 'core/domain/index';
+import { IPipelineValidationResults, PIPELINE_CONFIG_VALIDATOR, PipelineConfigValidator } from '../validation/pipelineConfig.validator';
+import { IExecutionViewState, IPipelineLink, IPipelineNode, PIPELINE_GRAPH_SERVICE, PipelineGraphService } from './pipelineGraph.service';
+
+require('./pipelineGraph.less');
+
+export class PipelineGraphController implements ng.IComponentController {
+  public pipeline: IPipeline;
+  public execution: IExecution;
+  public viewState: IExecutionViewState;
+  public onNodeClick: (node: IPipelineNode) => void;
+  public shouldValidate: boolean;
+
+  static get $inject() {
+    return ['$scope', '$element', '$', '$window', 'pipelineGraphService', 'pipelineConfigValidator'];
+  }
+
+  public constructor(private $scope: any,
+                     private $element: JQuery,
+                     private $: any,
+                     private $window: ng.IWindowService,
+                     private pipelineGraphService: PipelineGraphService,
+                     private pipelineConfigValidator: PipelineConfigValidator
+                     ) {}
+
+  public $onInit(): void {
+    let pipelineValidations: IPipelineValidationResults = { pipeline: [], stages: [] };
+
+    this.$scope.warningsPopover = require('./warnings.popover.html');
+
+    // track and save the graph scroll position for executions so it doesn't get reset to
+    // zero every second due to repaint.
+    this.$element.on('mousewheel', () => {
+      if (this.execution) {
+        this.pipelineGraphService.xScrollOffset[this.execution.id] = this.$element.scrollLeft();
+      }
+    });
+
+    const minLabelWidth = 100;
+
+    this.$scope.nodeRadius = 8;
+    this.$scope.rowPadding = 20;
+    this.$scope.graphVerticalPadding = 11;
+    this.$scope.minExecutionGraphHeight = 40;
+    this.$scope.labelOffsetX = this.$scope.nodeRadius + 3;
+    this.$scope.labelOffsetY = this.$scope.nodeRadius + 10;
+    let graphId = this.pipeline ? this.pipeline.id : this.execution.id;
+
+    /**
+     * Used to draw inverse bezier curve between stages
+     */
+    const curvedLink = (d: any) => {
```

Comment:
Any reason to not make all these types of functions either `public` or `private` as necessary?  I mean, if we're going to Typescript, let's go all the way.